### ISSUE TITLE
A load of deferredjs via AsyncFragmentLoader$HttpDownloadFailure returns 404

### ISSUE THREAD

**Issue by mparfianowicz** [gwtproject/gwt#9732]
**GWT version:** 2.9.0
**Browser (with version):** Firefox
**Operating System:** MacOS, Linux, Windows

---

##### Description

A load of a deferred JS code fails with the 'HTTP 404' error, despite the fact that the resource is present and on the server side GET operation returns HTTP 200.

A stack trace looks like this:

```Failed to load administration panel:com.google.gwt.core.client.impl.AsyncFragmentLoader$HttpDownloadFailure: Download of http://localhost:6060/xxxxxx/AdminGwtModule/deferredjs/F2F06143A0520BB4CF58E8F927992C9A/7.cache.js?autoRetry=3 failed with status 404(Script Tag Failure - no status available)
            Unknown.java_lang_Throwable_Throwable__Ljava_lang_String_2V(http://localhost:6060/xxxxxx/AdminGwtModule/F2F06143A0520BB4CF58E8F927992C9A.cache.js)
            Unknown.java_lang_Exception_Exception__Ljava_lang_String_2V(http://localhost:6060/xxxxxx/AdminGwtModule/F2F06143A0520BB4CF58E8F927992C9A.cache.js)
            Unknown.java_lang_RuntimeException_RuntimeException__Ljava_lang_String_2V(http://localhost:6060/xxxxxx/AdminGwtModule/F2F06143A0520BB4CF58E8F927992C9A.cache.js)
            Unknown.com_google_gwt_core_client_impl_AsyncFragmentLoader$HttpDownloadFailure_AsyncFragmentLoader$HttpDownloadFailure__Ljava_lang_String_2ILjava_lang_String_2V(http://localhost:6060/xxxxxx/AdminGwtModule/F2F06143A0520BB4CF58E8F927992C9A.cache.js)
            Unknown.com_google_gwt_core_client_impl_ScriptTagLoadingStrategy_cleanup__Lcom_google_gwt_core_client_impl_LoadingStrategyBase$RequestData_2V(http://localhost:6060/xxxxxx/AdminGwtModule/F2F06143A0520BB4CF58E8F927992C9A.cache.js)
            Unknown.com_google_gwt_core_client_impl_ScriptTagLoadingStrategy$ScriptTagDownloadStrategy$1_onSuccess__Ljava_lang_Void_2V(http://localhost:6060/xxxxxx/AdminGwtModule/F2F06143A0520BB4CF58E8F927992C9A.cache.js)
            Unknown.com_google_gwt_core_client_impl_ScriptTagLoadingStrategy$ScriptTagDownloadStrategy$1_onSuccess__Ljava_lang_Object_2V(http://localhost:6060/xxxxxx/AdminGwtModule/F2F06143A0520BB4CF58E8F927992C9A.cache.js)
            Unknown.onload<(http://localhost:6060/xxxxxx/AdminGwtModule/F2F06143A0520BB4CF58E8F927992C9A.cache.js)
            Unknown.com_google_gwt_core_client_impl_Impl_apply__Ljava_lang_Object_2Ljava_lang_Object_2Ljava_lang_Object_2Ljava_lang_Object_2(http://localhost:6060/xxxxxx/AdminGwtModule/F2F06143A0520BB4CF58E8F927992C9A.cache.js)
            Unknown.com_google_gwt_core_client_impl_Impl_entry0__Ljava_lang_Object_2Ljava_lang_Object_2Ljava_lang_Object_2Ljava_lang_Object_2(http://localhost:6060/xxxxxx/AdminGwtModule/F2F06143A0520BB4CF58E8F927992C9A.cache.js)
            Unknown.com_google_gwt_core_client_impl_Impl_entry__Lcom_google_gwt_core_client_JavaScriptObject_2Lcom_google_gwt_core_client_JavaScriptObject_2/<(http://localhost:6060/xxxxxx/AdminGwtModule/F2F06143A0520BB4CF58E8F927992C9A.cache.js)
            Unknown.anonymous(Unknown)```



##### Steps to reproduce

I have no specific steps to reproduce. This usually happens randomly in CI when a selenium/webdriver quickly clicks on links and pages or page fragments are being reloaded.


##### Links to further discussions

**Comment by mparfianowicz** [comment]
I might have found a root cause, I'm working on a pull request, please stay tuned for updates.

**Comment by mparfianowicz** [comment]
I created a [PR](https://github.com/atlassian-forks/gwt/pull/1) in an Atlassian fork. The change is that it no longer throws an exception from the onSuccess() call. 


I looked into code refactoring (made ~10 years ago) in this code fragment to see how it works now and how it worked before:

<img width="1515" alt="ScriptTagDownloadStrategy-refactoring-1" src="https://user-images.githubusercontent.com/1674007/122734668-93a97280-d27e-11eb-9cfc-254659a7fdda.png">

<img width="771" alt="ScriptTagDownloadStrategy-refactoring-2" src="https://user-images.githubusercontent.com/1674007/122734929-d4a18700-d27e-11eb-8723-d3d2e4d76b41.png">
<img width="1490" alt="ScriptTagDownloadStrategy-refactoring-3" src="https://user-images.githubusercontent.com/1674007/122734938-d703e100-d27e-11eb-9598-cd539d972781.png">


In the old version prior to refactoring, there were two separate callbacks for `onSuccess()` and `onFailure()` events.
The `onFailure()` callback deleted an entry from the hashmap and next called the `callOnLoadError()` handler, which was  was throwing the HttpDownloadFailure with HTTP 404 code. The `onSuccess()` callback deleted an entry from the hashmap too and next called the `tryInstall()` method.

In the new version both `onSuccess()` and onFailure()` callbacks call the same `cleanup()` method. The method deletes an entry from the hashmap and throws HttpDownloadFailure with 404 code. Where is the script installation? In the new version we call setAsyncCallback() at the beginning of the tryDownload method, which adds asyncCallback to the __gwtModuleFunction hashmap. And the asyncCallback calls the `tryInstall()` method.

**Comment by mparfianowicz** [comment]
Thus the question is why do we have any code in the `onSuccess()` callback at all? Intuition suggests that this method might be empty, actually. However, this JavaDoc comment in the ScriptInjector class is intriguing:

```
* On Safari version 3 and prior, the onSuccess() callback may be invoked
* even when the load of a page fails.  
...
public FromUrl setCallback(Callback<Void, Exception> callback) {
```

**Comment by mparfianowicz** [comment]
In my PR I decided to go a bit 'safe path' and instead of having an empty onSuccess() method, I delete an entry from a hashmap, but without throwing an exception.

Dear GWT maintainers, could you please review my analysis and a PR if the fix is correct? In case not, what solution do you suggest?

**Comment by tbroyer** [comment]
Rather than screenshots, the diff can be read at: https://github.com/gwtproject/gwt/commit/f6e46b88f954d7408994e4cf26c001b526e0a1c6#diff-ef1cccd118f1e769947fb24de22082ccef6288c67b07c5800adf11cb7b215a1b

You missed the fact that `setOnFailure` in the previous version *did* set an `onload` that would call the `onFailure` callback too. The code, to me, works exactly the same before and after that change, with the `tryInstall` called by the script itself (the CrossSiteIframeLinker wraps the deferred fragment such that the script, when loaded, will call the `__gwtModuleFunction['runAsyncCallback'+fragment]` function, that will clean itself; and this expects that this si done before the `onload` or `onerror` callbacks are called, which will then check if the function has been cleaned already, indicating a success, or not, indicating an error: either the script didn't load).
This code is subject to race conditions in case `onload` is triggered before the script itself is run, but that would be non-compliant: https://html.spec.whatwg.org/multipage/scripting.html#execute-the-script-block
The other possibility is that your script actually doesn't run because it itself fails for some reason (check the errors in the console).
Or maybe the script was blocked by CSP or similar (e.g. SRI), so the server correctly served it but the browser refused to run it as it violated some policies.

Fwiw, I won't look at your PR until you sign the CLA, but if you only removed the call to `cleanup` from `onSuccess`, then it's the wrong fix anyway.

**Comment by mparfianowicz** [comment]
Hi Thomas, thank you for your quick reply. 

>  I won't look at your PR until you sign the CLA

I'm working on it, this will be a company CLA (Atlassian).

> If you only removed the call to cleanup from onSuccess, then it's the wrong fix anyway.

Thank you for clarification. I will look at these different possibilities listed by you.

Cheers, Marek
